### PR TITLE
#34 - Go SDK - Add SafeGet, SafeSet, SafeReference, SafeAdd, Backup and Restore

### PR DESCRIPTION
**Pull request by padurean**

### CODE REVIEW

File: pkg/client/client.go [hunk 1]
```diff
@@ -86,6 +89,56 @@ func (c *ImmuClient) Get(keyReader io.Reader) (*schema.Item, error) {
...
+	var metadata runtime.ServerMetadata
+	safeItem, err := c.serviceClient.SafeGet(
+		context.Background(),
+		opts,
+		grpc.Header(&metadata.HeaderMD),
+		grpc.Trailer(&metadata.TrailerMD))
+
+	verified := safeItem.Proof.Verify(safeItem.Item.Hash(), *root)
+	if verified {
+		// saving a fresh root
```
Comment (padurean) [inline pkg/client/client.go]: @mmeloni Do we need to do this or not?

File: pkg/client/client.go [hunk 2]
```diff
@@ -86,6 +89,56 @@ func (c *ImmuClient) Get(keyReader io.Reader) (*schema.Item, error) {
 	return result, err
 }
 
+// SafeGet ...
+func (c *ImmuClient) SafeGet(keyReader io.Reader) (*schema.Item, bool, error) {
```
Comment (padurean) [inline pkg/client/client.go]: @mmeloni Do you think it is ok to return an `Item` or should we create a dedicated struct (for simple Get an `Item` is returned, so probably if we do not want to expose it we need to change everywhere in this file).

File: pkg/client/client.go [hunk 3]
```diff
@@ -86,6 +91,70 @@ func (c *ImmuClient) Get(keyReader io.Reader) (*schema.Item, error) {
...
+	start := time.Now()
+	if !c.isConnected() {
+		return nil, ErrNotConnected
+	}
+	key, err := ioutil.ReadAll(keyReader)
+	if err != nil {
+		return nil, err
+	}
+
+	rs := NewRootService(c.serviceClient, cache.NewFileCache())
```
Comment (mmeloni) [inline pkg/client/client.go:113]: Can you use dependency injection for RootService?

File: pkg/client/client.go [hunk 4]
```diff
@@ -245,6 +403,71 @@ func (c *ImmuClient) Reference(keyReader io.Reader, valueReader io.Reader) (*sch
...
+	root, err := rs.GetRoot(context.Background())
+	if err != nil {
+		return nil, err
+	}
+
+	opts := &schema.SafeReferenceOptions{
+		Ro: &schema.ReferenceOptions{
+			Reference: &schema.Key{Key: reference},
```
Comment (mmeloni) [inline pkg/client/client.go:429]: In the future We should semplify this model inside .proto file and use a more simple
```
message ReferenceOptions {
	bytes reference = 1;
	bytes key = 2;
}
```

File: pkg/client/client.go (hunk 3 above)
Comment (padurean) [reply pkg/client/client.go]: @mmeloni Please see my last commit : https://github.com/codenotary/immudb/pull/40/commits/fb49cb630987c7e5178f8b83ee146fdf6cedb461

and let me know what you think. Thanks.

Comment (mmeloni) [reply pkg/client/client.go]: ok!

**Pull request closed without merging**